osu.Server.BeatmapSubmission/Services/BeatmapPackagePatcher.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Extensions;
+using osu.Game.Beatmaps.Formats;
+using osu.Game.IO;
+using osu.Game.IO.Archives;
+using SharpCompress.Common;
+using SharpCompress.Writers;
+using SharpCompress.Writers.Zip;
+
+namespace osu.Server.BeatmapSubmission.Services
+{
+    public class BeatmapPackagePatcher
+    {
+        public static readonly ZipWriterOptions DEFAULT_ZIP_WRITER_OPTIONS = new ZipWriterOptions(CompressionType.Deflate)
+        {
+            ArchiveEncoding = ZipArchiveReader.DEFAULT_ENCODING,
+        };
+
+        private readonly IBeatmapStorage beatmapStorage;
+
+        public BeatmapPackagePatcher(IBeatmapStorage beatmapStorage)
+        {
+            this.beatmapStorage = beatmapStorage;
+        }
+
+        public async Task<MemoryStream> PatchBeatmapSetAsync(
```

Comment:
I'm wondering if we want some way of confirming that something actually changed, to avoid creating new versions server-side if there weren't any changes. Or do we just trust that the client does this correctly?

Reply:
Can agree with the sentiment, but I'd see it as a follow-up change for everyone's sake. Will put a TODO in for now.